kahypar/partition/bin_packing/prepacking.h

```diff
@@ -0,0 +1,186 @@
+  const std::vector<HypernodeID> nodes = nodesInDescendingWeightOrder(hg);
+  TwoLevelPacker<BPAlgorithm> packer(rb_range_k, max_bin_weight);
+  std::vector<std::pair<HypernodeWeight, HypernodeWeight>> weights;
+  weights.reserve(nodes.size() + 1);
+  HypernodeWeight sum = 0;
+  for (size_t i = 0; i < nodes.size(); ++i) {
+    HypernodeWeight w = hg.nodeWeight(nodes[i]);
+    weights.emplace_back(w, sum);
+    sum += w;
+  }
+  weights.emplace_back(0, sum);
+  const PartitionID max_k = *std::max_element(num_bins_per_part.cbegin(), num_bins_per_part.cend());
+  const BalanceSegTree seg_tree(weights, max_k);
+  std::vector<PartitionID> parts;
+
+  ASSERT(nodes.size() > 0);
+  ASSERT([&]() {
+    for (size_t i = 0; i < static_cast<size_t>(context.partition.k); ++i) {
+      if (upper_weight[i] < context.initial_partitioning.perfect_balance_partition_weight[i]
+        || upper_weight[i] >= num_bins_per_part[i] * max_bin_weight
+        ) {
+        return false;
+      }
+    }
+    return true;
+  } (), "The allowed partition or bin weights are to small to find a valid partition.");
+
+  size_t i;
+  std::pair<PartitionMapping, std::vector<HypernodeWeight>> packing_result = packer.applySecondLevel(upper_weight, num_bins_per_part);
+  for (i = 0; i < nodes.size(); ++i) {
+    ASSERT(upper_weight.size() == packing_result.second.size());
+
+    const size_t max_part_idx = getMaxPartIndex(context, packing_result.second, weights[i].first, true, max_bin_weight);
+    const HypernodeWeight remaining_weight_for_max_part = upper_weight[max_part_idx] - packing_result.second[max_part_idx];
+    const HypernodeWeight searched_weight = weights[i].second + remaining_weight_for_max_part;
+
+    // find subrange of specified weight
+    const size_t j = std::lower_bound(weights.cbegin() + i, weights.cend(), std::make_pair(0, searched_weight),
+                     [&](const auto& v1, const auto& v2) {
+                       return v1.second < v2.second;
+                     }) - weights.cbegin();
+
+    // calculate upper bound for the additional weight required by vertices that are not prepacked
+    if (j > i) {
+      const HypernodeWeight upper_bound = seg_tree.query(i, j - 1) - weights[i].second;
+      if (packing_result.second[max_part_idx] + upper_bound <= num_bins_per_part[max_part_idx] * max_bin_weight) {
+        break;
+      }
+    }
+
+    parts.push_back(packer.insertElement(weights[i].first));
+    packing_result = packer.applySecondLevel(upper_weight, num_bins_per_part);
+  }
+  ASSERT(parts.size() <= nodes.size());
+
+  packing_result.first.applyMapping(parts);
+  for (size_t i = 0; i < parts.size(); ++i) {
+    hg.setFixedVertex(nodes[i], parts[i]);
+  }
+}
+
+template< class BPAlgorithm >
+static inline void calculateHeuristicPrepacking(Hypergraph& hg, const Context& context, const PartitionID& rb_range_k, const HypernodeWeight max_bin_weight) {
```

Comment:
Do we still use both `calculateHeuristicPrepacking` and `calculateHeuristicPrepacking`?

Reply:
The heuristic prepacking is not used in the paper, but I think it might make sense to keep it. The reason we do not use it in the paper is to keep things simpler in the theoretical description. Also we ran some preliminary experiments, with the result that using the heuristic prepacking improves the result quality, but the difference was quite small so it is no problem to exclude it in the paper.

I have updated the code now to include an additional parameter `i-bp-heuristic-prepacking` which enables/disables the heuristic prepacking (which is the same I did for the experiments). For the current configs I disabled it to match the version of the paper.